download/upgradeRecipe/upgradeRecipe8.adoc

```diff
@@ -1577,3 +1577,16 @@ After in `*StatefulVariableListener.java*`:
   }
 
 ----
+
+== From 8.2.0.Final to 8.3.0.Final
+
+[.upgrade-recipe-minor.upgrade-recipe-constraint-match-equals]
+=== `ConstraintMatch.compareTo()` inconsistent with `equals()`
+
+The `equals()` override in `ConstraintMatch` has been removed.
+As a result, no two different instances of the class will be considered equal.
+This is in contrast to the `compareTo()` method, which continues to consider two instances equal
+if all their field values are equal.
+
+Behavior of `ConstraintMatch` has always been considered an implementation detail.
+Users are discouraged from relying on undocumented behavior.
```

Comment:
I'd remove these 2 last to lines. It's too defensive and it's out of place.
We could use some motivation for why we're doing this.

How about: 
The `equals()` override in `ConstraintMatch` has been removed to allow for more advanced ConstraintStream operations that can't guarantee distinctness.

Or:
The `equals()` override in `ConstraintMatch` has been removed to not fail on constraints with non-distinct matches.